### ISSUE TITLE
Flaky tests: `ImageList` randomly crashes tests and result in AccessViolationException

### ISSUE THREAD

**Issue by RussKie** [dotnet/winforms#3358]
For the past few days I'm observing tests randomly fail in Release builds, with some builds providing **no logs** of failures.
Few other logs I found are below:

```
    System.Windows.Forms.Tests.ImageCollectionTests.ImageCollection_IListItem_Set_GetReturnsExpected(transparentColor: Color [Empty], value: Bitmap { Flags = 2, FrameDimensionsList = [7462dc86-6180-4c7e-8e3f-ee7333a7a483], Height = 256, HorizontalResolution = 96, Palette = ColorPalette { Entries = [...], Flags = 7929940 }, ... }) [FAIL]
      System.InvalidOperationException : Image cannot be added to the ImageList.
      Stack Trace:
        /_/src/System.Windows.Forms/src/System/Windows/Forms/ImageList.cs(414,0): at System.Windows.Forms.ImageList.AddToHandle(Bitmap bitmap)
        /_/src/System.Windows.Forms/src/System/Windows/Forms/ImageList.cs(485,0): at System.Windows.Forms.ImageList.CreateHandle()
        /_/src/System.Windows.Forms/src/System/Windows/Forms/ImageList.cs(142,0): at System.Windows.Forms.ImageList.get_Handle()
        /_/src/System.Windows.Forms.Primitives/src/Interop/ComCtl32/Interop.ImageList.Replace.cs(20,0): at Interop.ComCtl32.ImageList.Replace(IHandle himl, Int32 i, IntPtr hbmImage, IntPtr hbmMask)
        /_/src/System.Windows.Forms/src/System/Windows/Forms/ImageList.ImageCollection.cs(166,0): at System.Windows.Forms.ImageList.ImageCollection.set_Item(Int32 index, Image value)
        /_/src/System.Windows.Forms/src/System/Windows/Forms/ImageList.ImageCollection.cs(199,0): at System.Windows.Forms.ImageList.ImageCollection.System.Collections.IList.set_Item(Int32 index, Object value)
        /_/src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ImageList.ImageCollectionTests.cs(1626,0): at System.Windows.Forms.Tests.ImageCollectionTests.ImageCollection_IListItem_Set_GetReturnsExpected(Color transparentColor, Image value)
```

```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Repeat 2 times:
--------------------------------
   at Interop+User32.CallWindowProcW(IntPtr, IntPtr, WM, IntPtr, IntPtr)
--------------------------------
   at System.Windows.Forms.NativeWindow.DefWndProc(System.Windows.Forms.Message ByRef)
   at System.Windows.Forms.Control.DefWndProc(System.Windows.Forms.Message ByRef)
   at System.Windows.Forms.Control.WndProc(System.Windows.Forms.Message ByRef)
   at System.Windows.Forms.ListView.WndProc(System.Windows.Forms.Message ByRef)
   at System.Windows.Forms.Control+ControlNativeWindow.OnMessage(System.Windows.Forms.Message ByRef)
   at System.Windows.Forms.Control+ControlNativeWindow.WndProc(System.Windows.Forms.Message ByRef)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr, WM, IntPtr, IntPtr)
   at Interop+User32.SendMessageW(IntPtr, WM, IntPtr, IntPtr)
   at Interop+User32.SendMessageW(IntPtr, WM, IntPtr, IntPtr)
   at Interop+User32.SendMessageW(IHandle, WM, IntPtr, IntPtr)
   at System.Windows.Forms.ListView.UpdateExtendedStyles()
   at System.Windows.Forms.ListView.OnHandleCreated(System.EventArgs)
   at System.Windows.Forms.Control.WmCreate(System.Windows.Forms.Message ByRef)
   at System.Windows.Forms.Control.WndProc(System.Windows.Forms.Message ByRef)
   at System.Windows.Forms.ListView.WndProc(System.Windows.Forms.Message ByRef)
   at System.Windows.Forms.Control+ControlNativeWindow.OnMessage(System.Windows.Forms.Message ByRef)
   at System.Windows.Forms.Control+ControlNativeWindow.WndProc(System.Windows.Forms.Message ByRef)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr, WM, IntPtr, IntPtr)
   at Interop+User32.CreateWindowExW(WS_EX, Char*, System.String, WS, Int32, Int32, Int32, Int32, IntPtr, IntPtr, IntPtr, System.Object)
   at Interop+User32.CreateWindowExW(WS_EX, Char*, System.String, WS, Int32, Int32, Int32, Int32, IntPtr, IntPtr, IntPtr, System.Object)
   at Interop+User32.CreateWindowExW(WS_EX, System.String, System.String, WS, Int32, Int32, Int32, Int32, IntPtr, IntPtr, IntPtr, System.Object)
   at System.Windows.Forms.NativeWindow.CreateHandle(System.Windows.Forms.CreateParams)
   at System.Windows.Forms.Control.CreateHandle()
   at System.Windows.Forms.ListView.CreateHandle()
   at System.Windows.Forms.Control.get_Handle()
   at System.Windows.Forms.Tests.ListViewTests.ListView_StateImageList_SetWithHandle_GetReturnsExpected(Boolean, Boolean, Boolean, Boolean, System.Windows.Forms.View, System.Windows.Forms.ImageList, Int32, Int32)
   at System.RuntimeMethodHandle.InvokeMethod(System.Object, System.Object[], System.Signature, Boolean, Boolean)
   at System.Reflection.RuntimeMethodInfo.Invoke(System.Object, System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo)
   at System.Reflection.MethodBase.Invoke(System.Object, System.Object[])
   at Xunit.Sdk.TestInvoker`1[[System.__Canon, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].CallTestMethod(System.Object)
   at Xunit.Sdk.UITestInvoker+<>c__DisplayClass2_0+<<RunAsync>b__2>d.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Xunit.Sdk.UITestInvoker+<>c__DisplayClass2_0+<<RunAsync>b__2>d, Xunit.StaFact, Version=1.0.0.0, Culture=neutral, PublicKeyToken=593f35978b459a4b]].ExecutionContextCallback(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Xunit.Sdk.UITestInvoker+<>c__DisplayClass2_0+<<RunAsync>b__2>d, Xunit.StaFact, Version=1.0.0.0, Culture=neutral, PublicKeyToken=593f35978b459a4b]].MoveNext(System.Threading.Thread)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[Xunit.Sdk.UITestInvoker+<>c__DisplayClass2_0+<<RunAsync>b__2>d, Xunit.StaFact, Version=1.0.0.0, Culture=neutral, PublicKeyToken=593f35978b459a4b]].MoveNext()
   at Xunit.Sdk.Utilities+SyncContextAwaiter+<>c.<OnCompleted>b__5_0(System.Object)
   at System.RuntimeMethodHandle.InvokeMethod(System.Object, System.Object[], System.Signature, Boolean, Boolean)
   at System.Reflection.RuntimeMethodInfo.Invoke(System.Object, System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo)
   at System.Delegate.DynamicInvokeImpl(System.Object[])
   at System.Delegate.DynamicInvoke(System.Object[])
   at System.Windows.Forms.Control.InvokeMarshaledCallbackDo(ThreadMethodEntry)
   at System.Windows.Forms.Control.InvokeMarshaledCallbackHelper(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Windows.Forms.Control.InvokeMarshaledCallback(ThreadMethodEntry)
   at System.Windows.Forms.Control.InvokeMarshaledCallbacks()
   at System.Windows.Forms.Control.WndProc(System.Windows.Forms.Message ByRef)
   at System.Windows.Forms.Control+ControlNativeWindow.OnMessage(System.Windows.Forms.Message ByRef)
   at System.Windows.Forms.Control+ControlNativeWindow.WndProc(System.Windows.Forms.Message ByRef)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr, WM, IntPtr, IntPtr)
   at Interop+User32.DispatchMessageW(MSG ByRef)
   at Interop+User32.DispatchMessageW(MSG ByRef)
   at System.Windows.Forms.Application+ComponentManager.Interop.Mso.IMsoComponentManager.FPushMessageLoop(UIntPtr, msoloop, Void*)
   at System.Windows.Forms.Application+ThreadContext.RunMessageLoopInner(msoloop, System.Windows.Forms.ApplicationContext)
   at System.Windows.Forms.Application+ThreadContext.RunMessageLoop(msoloop, System.Windows.Forms.ApplicationContext)
   at Xunit.Sdk.WinFormsSynchronizationContextAdapter.PumpTill(System.Threading.SynchronizationContext, System.Threading.Tasks.Task)
   at Xunit.Sdk.ThreadRental+<>c__DisplayClass11_0.<CreateAsync>b__0()
   at System.Threading.ThreadHelper.ThreadStart_Context(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.ThreadHelper.ThreadStart()
```

**Comment by weltkante** [comment]
If you want to get minidumps on crashes there is a registry setting for that, but its probably all just AccessViolations after some memory corruption so the stack traces probably won't be too useful. (If an AccessViolation happens outside managed code then it just terminates the process.)

**Comment by RussKie** [comment]
I'll try to get some time to plugin the screenshot capture and a memory dump functionality into the CI pipeline.

**Comment by weltkante** [comment]
Not getting any AVs or process crashes locally.

**Comment by RussKie** [comment]
Some more: https://dev.azure.com/dnceng/public/_build/results?buildId=662467&view=artifacts&type=publishedArtifacts
```
    System.Windows.Forms.Tests.ImageCollectionTests.ImageCollection_AddStrip_InvokeWithHandle_Success(transparentColor: Color [Black], value: Bitmap { Flags = 2, FrameDimensionsList = [7462dc86-6180-4c7e-8e3f-ee7333a7a483], Height = 16, HorizontalResolution = 96, Palette = ColorPalette { Entries = [...], Flags = 85653360 }, ... }, expectedCount: 1) [FAIL]
      System.InvalidOperationException : Image cannot be added to the ImageList.
      Stack Trace:
        F:\workspace\_work\1\s\src\System.Windows.Forms\src\System\Windows\Forms\ImageList.cs(414,0): at System.Windows.Forms.ImageList.AddToHandle(Bitmap bitmap)
        F:\workspace\_work\1\s\src\System.Windows.Forms\src\System\Windows\Forms\ImageList.ImageCollection.cs(336,0): at System.Windows.Forms.ImageList.ImageCollection.Add(Original original, ImageInfo imageInfo)
        F:\workspace\_work\1\s\src\System.Windows.Forms\src\System\Windows\Forms\ImageList.ImageCollection.cs(427,0): at System.Windows.Forms.ImageList.ImageCollection.AddStrip(Image value)
        F:\workspace\_work\1\s\src\System.Windows.Forms\tests\UnitTests\System\Windows\Forms\ImageList.ImageCollectionTests.cs(881,0): at System.Windows.Forms.Tests.ImageCollectionTests.ImageCollection_AddStrip_InvokeWithHandle_Success(Color transparentColor, Image value, Int32 expectedCount)
```

One thing I didn't notice before, only x86 tests appear to fail, which suggests bitness-related issues.

**Comment by RussKie** [comment]
Another IOE in Release:
```
    System.Windows.Forms.Tests.ImageCollectionTests.ImageCollection_AddStrip_InvokeWithHandle_Success(transparentColor: Color [Black], value: Bitmap { Flags = 2, FrameDimensionsList = [7462dc86-6180-4c7e-8e3f-ee7333a7a483], Height = 16, HorizontalResolution = 96, Palette = ColorPalette { Entries = [...], Flags = 6619213 }, ... }, expectedCount: 2) [FAIL]
      System.InvalidOperationException : Image cannot be added to the ImageList.
      Stack Trace:
        /_/src/System.Windows.Forms/src/System/Windows/Forms/ImageList.cs(414,0): at System.Windows.Forms.ImageList.AddToHandle(Bitmap bitmap)
        /_/src/System.Windows.Forms/src/System/Windows/Forms/ImageList.ImageCollection.cs(336,0): at System.Windows.Forms.ImageList.ImageCollection.Add(Original original, ImageInfo imageInfo)
        /_/src/System.Windows.Forms/src/System/Windows/Forms/ImageList.ImageCollection.cs(427,0): at System.Windows.Forms.ImageList.ImageCollection.AddStrip(Image value)
        /_/src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ImageList.ImageCollectionTests.cs(857,0): at System.Windows.Forms.Tests.ImageCollectionTests.ImageCollection_AddStrip_InvokeWithHandle_Success(Color transparentColor, Image value, Int32 expectedCount)
```
https://dev.azure.com/dnceng/public/_build/results?buildId=672867&view=results

**Comment by weltkante** [comment]
I had that same `ImageCollection_AddStrip_InvokeWithHandle_Success` test fail locally reporting that `AddStrip` loaded zero images (expected one image)

These failures make no sense, makes memory corruption the most likely cause.

**Comment by hughbe** [comment]
Also failure in `ImageCollection_IListItem_SetWithHandle_GetReturnsExpected`
```
System.InvalidOperationException : Image cannot be added to the ImageList.
   at System.Windows.Forms.ImageList.AddToHandle(Bitmap bitmap) in /_/src/System.Windows.Forms/src/System/Windows/Forms/ImageList.cs:line 414
   at System.Windows.Forms.ImageList.CreateHandle() in /_/src/System.Windows.Forms/src/System/Windows/Forms/ImageList.cs:line 485
   at System.Windows.Forms.ImageList.get_Handle() in /_/src/System.Windows.Forms/src/System/Windows/Forms/ImageList.cs:line 142
   at System.Windows.Forms.Tests.ImageCollectionTests.ImageCollection_IListItem_SetWithHandle_GetReturnsExpected(Color transparentColor, Image value) in /_/src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ImageList.ImageCollectionTests.cs:line 1705
```

**Comment by RussKie** [comment]
I can't seem to repro this issue locally...
Looking through the code all tests appear to be failing at the same point - `System.Windows.Forms.ImageList.AddToHandle(Bitmap bitmap) in /_/src/System.Windows.Forms/src/System/Windows/Forms/ImageList.cs:line 414`.

I think it fails here:
https://github.com/dotnet/winforms/blob/d4cfc2d9d6cb20984761fc0cf5eb8efd70c2d810/src/System.Windows.Forms/src/System/Windows/Forms/ImageList.cs#L402-L415

If I found the right code for [`ImageList_Add`](https://docs.microsoft.com/windows/win32/api/commctrl/nf-commctrl-imagelist_add) then it's failing while invoking [`HIMAGELIST_QueryInterface`](https://docs.microsoft.com/windows/win32/api/commctrl/nf-commctrl-himagelist_queryinterface) on the supplied imagelist, and failing returns `-1`.

**Comment by hughbe** [comment]
This is really weird. I know its not related to your theory here, but could @weltkante's comment here https://github.com/dotnet/winforms/pull/3324#discussion_r429144722 cause this?

Perhaps there's some odd/incorrect/overlapping memory copying happening?

**Comment by weltkante** [comment]
I really doubt its a bug in the ImageList code itself, QueryInterface failing would mean the object is corrupted (playing into the memory corruption idea) or already released (which I'd expect to be handled gracefully).

>  but could @weltkante's comment here #3324 (comment) cause this?
> 
> Perhaps there's some odd/incorrect/overlapping memory copying happening?

Sure, you can cause arbitrary damage by bad memory copying. But I don't think that particular linked comment was right, later in the discussion it became clear that the caller made/provided a lot of assumptions that should have made the code safe.

> I can't seem to repro this issue locally

I had some locally, but very rare (one in several dozen runs), and never under a debugger. I think the best chance is getting some dumps (probably need a full dump though not a minidump). I'll look into what options we have, I don't think we can set the usual machine wide setting to capture dumps on process crash, so might need some native DLL stub to write out the dump, or a secondary helper process.

**Comment by RussKie** [comment]
@MattGal we seem to have "progressed" to level 2 of test nightmares, tests randomly fail with no clear indication of the underlying cause.
Is [collecting user-mode memory dumps](https://docs.microsoft.com/en-us/windows/win32/wer/collecting-user-mode-dumps) a possibility on CI agents (it requires HKLM registry settings)?

**Comment by RussKie** [comment]
Looking at a log for a failed build in https://github.com/dotnet/winforms/pull/3224#issuecomment-641226644 it appears we may have a a minidump (snip):
```
      Timed out at 6/9/2020 11:12:41 AM after 60000ms waiting for remote process.
      Wrote mini dump to: F:\h\w\B5E80973\w\AB98096F\uploads\5272.aw55rizx.gmv.dmp
      	Process ID: 5272
      	Handle: 10236
      	Name: dotnet
      	MainModule: F:\workspace\_work\1\s\.dotnet\dotnet.exe
      	Threads:
      		Thread #1 (OS 0x868)   
      			[InlinedCallFrame]
      			[InlinedCallFrame]
      			System.Environment.Exit(Int32)
      			Microsoft.DotNet.RemoteExecutor.Program.Main(System.String[])
      		Thread #2 (OS 0x16CC) [Finalizer] [Background] 
      			(...snip...)
      		Thread #3 (OS 0x12F8) [Thread pool worker] [Background] [MTA]
      		Thread #5 (OS 0x1A94)  [Background] 
      			(...snip...)
      		Thread #4 (OS 0x7CB4) [Thread pool worker] [Background] [MTA]
      		Thread #6 (OS 0x8BF4) [Thread pool worker] [Background] [MTA]
      
      Stack Trace:
        /_/src/Microsoft.DotNet.RemoteExecutor/src/RemoteInvokeHandle.cs(219,0): at Microsoft.DotNet.RemoteExecutor.RemoteInvokeHandle.Dispose(Boolean disposing)
        /_/src/Microsoft.DotNet.RemoteExecutor/src/RemoteInvokeHandle.cs(54,0): at Microsoft.DotNet.RemoteExecutor.RemoteInvokeHandle.Dispose()
        F:\workspace\_work\1\s\src\System.Windows.Forms\tests\UnitTests\AccessibleObjects\ListViewGroupAccessibleObjectTests.cs(142,0): at System.Windows.Forms.Tests.AccessibleObjects.ListViewGroupAccessibleObjectTests.ListViewGroupAccessibleObject_Bounds_ReturnsCorrectValue()
```

A question is whether we get a mini dump when a test runner crashes.

**Comment by hughbe** [comment]
> Microsoft.DotNet.RemoteExecutor.Program.Main

I thought `RemoteExecutor` is disabled in winforms?

**Comment by RussKie** [comment]
> I thought `RemoteExecutor` is disabled in winforms?

We disabled it, but this is a new test being developed that hasn't been skipped.
I've actually wondered whether we could re-enable RE, but apparently not :(

**Comment by weltkante** [comment]
Do you have access to it? Its not in the artifact zip you can publically download (it just has log and binlog files)

**Comment by weltkante** [comment]
> A question is whether we get a mini dump when a test runner crashes.

no

**Comment by weltkante** [comment]
Turned on the dump collection flags locally and have been repeating testruns for a few hours, but I got a [full crash dump](https://1drv.ms/u/s!AuW3VU_s-jxTgeZj9YzaRb7aZlSRCw?e=0cBE1T) now. No entry in log but the crashing test is `List​View_State​Image​List_Set​With​Handle​With​Non​Null​Old​Value_Get​Returns​Expected`, panic crash due to memory corruption, so yeah my guess was probably right.

**Comment by MattGal** [comment]
> @MattGal we seem to have "progressed" to level 2 of test nightmares, tests randomly fail with no clear indication of the underlying cause.
> Is [collecting user-mode memory dumps](https://docs.microsoft.com/en-us/windows/win32/wer/collecting-user-mode-dumps) a possibility on CI agents (it requires HKLM registry settings)?

Not only is it a possibility, we've already done it for you!  In fact for the run in question I think [this is a dump... doesn't seem to have your specific failure?](https://helixri107v0xdeko0k025g8.blob.core.windows.net/results-d76a85fe0f0d4801bb/5cf69245-042e-412f-965e-09ac971484cf/dotnet.exe.3284.dmp?sv=2019-02-02&se=2020-09-07T12%3A45%3A18Z&sr=c&sp=rl&sig=hX4iMcRy4B6Nnrt%2Boa9UaauEyW1%2FEOvzY6DhXxHZCzg%3D) where the exception is "The thread tried to read from or write to a virtual address for which it does not have the appropriate access" and from looking at the various threads running in this dump I wonder if it's just that you're stressing windows with a ton of native pinvokes that may not be intended to be called concurrently in the same process.  Just an idea, it's been a long time since I've been on the Windows team :)

I've got two ideas to try to help you.  Note I haven't done much more than review your code and builds and open up the dmp file above and looked at the AV in it.

- First, I've asked @marklio over teams who is kind of a PInvoke whisperer to look at your code just in case (there's something fishy about the problem being release specific).  No guarantee he'll have time to help and no promises but it might be something he has context or contacts on. 

- Secondly, I think given you are strongly asserting that this never repros locally that the most likely difference between your computer and the build machines is that they're Server Windows 10 SKUs and you use client, ergo various memory management things are different and the native windows things you're calling into might vary too.  So, while it's definitely worth understanding what's going on here (classically there are always various devs and power users who insist on running server SKU as their development environment) you can probably make the flaky CI problem go away (but not whatever this underlying bug is!) by changing `buildpool.server.amd64.vs2019.open` to `buildpool.windows.10.amd64.vs2019.open` .  If this unblocks you, please don't forget that you very likely have a real Windows Server bug and there are always customers quirky enough to use Winforms there.

Finally, I can (if you can get a corpnet jump box or use mine) provide you with a machine running our Server 2019 image, though I expect simply using the latest version from the Azure Gallery is likely to be able to fail the same way these build machines do, if that turns out to be the problem.

**Comment by hughbe** [comment]
> First, I've asked @marklio over teams who is kind of a PInvoke whisperer to look at your code just in case (there's something fishy about the problem being release specific). No guarantee he'll have time to help and no promises but it might be something he has context or contacts on.

I'm not sure if this is release specific - see crash in https://dev.azure.com/dnceng/public/_build/results?buildId=678166&view=logs&jobId=1f132584-ab08-5fbf-e5d0-5685de454342

> System​.Windows​.Forms​.Tests​.WebBrowserTests​.WebBrowser​_Refresh​_Invoke​Web​Browser​Refresh​Option​With​Document​Text​_Success
System​.Windows​.Forms​.Tests​.Html​Document​Tests​.Html​Document​_Link​Color​_Get​Custom​Value​On​Body​_Returns​Expected
all other are are either starting up the thread or waiting to create their first handle, so most likely have not performed any interop yet

This is interesting because these tests are (relatively) new. There be a chance that bad interop definitions could be causing this. For example we did see that `IHtmlElement4.FireEvent` was defined incorrectly in a previous PR

**Comment by weltkante** [comment]
> Not only is it a possibility, we've already done it for you!

@RussKie might want to learn where to find those ;)

> given you are strongly asserting that this never repros locally

I think thats a misunderstanding, its just failing very unreliably. Took me hours of local test runs to fail once with dumps enabled, but then it failed twice in a row (second dump not yet uploaded)

> there's something fishy about the problem being release specific

Definitely not release-only, I had it 3 times in debug build test runs so far, one of them being the dump I linked above.

> from looking at the various threads running in this dump

Its easy to misread this, but really there isn't that much going on. There is a lot of congestion at thread startup/shutdown and there is no throttling so it looks like a lot of tests are running in parallel, but they really aren't, most are still starting up/winding down.

Theres certainly potential for optimizing by e.g. throttling the number of UI threads running at a time to the number of cores or something, but I'd rather first fix the memory corruption while we still can reproduce it.

> In fact for the run in question I think this is a dump... doesn't seem to have your specific failure?

There is no specific failure, these are random test failures, most likely memory corruption.

In fact the crashing thread in the dump you linked seems to have is stack corrupted.

---

So far I've seen three dumps (the one you linked, the one I uploaded, and another one I made which I didn't upload yet sinve I've been busy). I guess I've been lucky that my first dump had so little active threads, the other two are much more work to look through. The three active tests (WebBrowser, HtmlDocument, ImageLists related) are all in common in the dumps so that doesn't help so far.

I agree with @hughbe that WebBrowser/Html code could have made a mistake during interop refactoring and are worth looking at the changesets again. Unfortunately WebBrowser engine is one of the hugest infrastructures in WinForms (it includes ActiveX hosting) so might not be easy to spot by review alone.

I'll try to build reduced test suites which just include the set of 3 test classes identified above and see if they corrupt memory if run repeatedly, maybe that can narrow down what we have to look at.

Of course feel free to run your own investigations if you have the time to do so, just wanted to let you know I'm on it.

**Comment by MattGal** [comment]
> > Not only is it a possibility, we've already done it for you!
> 
> @RussKie might want to learn where to find those ;)

Since these are builds you'll have to get Kusto access and use the AzDO API to figure out the "orchestration id" of your job to get this working.  It's a bit tricky (it'd be free if your tests ran on helix test machines).  @RussKie let me know if you want to send some instructions (and whether you have Kusto read access to engsrvprod).

**Comment by RussKie** [comment]
Matt showed me some magic tricks to look up dumps. Now we need to wait for the "right" memory dump to come along.

**Comment by RussKie** [comment]
I've downloaded the memory dump from https://dev.azure.com/dnceng/public/_build/results?buildId=678965&view=results
I'm seeing very much the same what @weltkante saw earlier in his dump - dozens (if not hundreds) of threads - most of which are queued up to create a handle, quite a few threads are pumping messages before their exits.


The immediate crash is a result of x-thread access:
![image](https://user-images.githubusercontent.com/4403806/84222558-39c68680-ab1b-11ea-9e7b-01478b894f9e.png)

The stack trace:
```
   at System.Windows.Forms.Control.MarshaledInvoke(Control caller, Delegate method, Object[] args, Boolean synchronous) in /_/src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:line 6870
   at System.Windows.Forms.Control.BeginInvoke(Delegate method, Object[] args) in /_/src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:line 4685
   at System.Windows.Forms.WindowsFormsSynchronizationContext.Post(SendOrPostCallback d, Object state) in /_/src/System.Windows.Forms/src/System/Windows/Forms/WindowsFormsSynchronizationContext.cs:line 91
   at Xunit.Sdk.Utilities.SyncContextAwaiter.OnCompleted(Action continuation) in D:\a\1\s\src\Xunit.StaFact\Sdk\Utilities.cs:line 34
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AwaitUnsafeOnCompleted[TAwaiter](TAwaiter& awaiter, IAsyncStateMachineBox box) in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs:line 149
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() in /_/src/libraries/System.Private.CoreLib/src/System/Runtime/ExceptionServices/ExceptionDispatchInfo.cs:line 57
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__140_1(Object state) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:line 1899
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:line 853
   at System.Threading.ThreadPoolWorkQueue.Dispatch() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:line 641
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback() in /_/src/coreclr/src/System.Private.CoreLib/src/System/Threading/ThreadPool.CoreCLR.cs:line 29
```

But it's unclear what other thread has led to this state...

**Comment by weltkante** [comment]
I doubt this is connected to the memory corruption, unless it happened to corrupt a SynchronizationContext. Trying to figure out which thread/task was failing in #3297 is a different investigation. It may be possible to figure out the originating thread/task by inspecting stack variables or heap objects. This doesn't help for the memory corruption investigation though.

**Comment by RussKie** [comment]
After merging #3423, which built successfully, the master build has failed - https://dev.azure.com/dnceng/public/_build/results?buildId=682086&view=results

I've downloaded the memdump, and it failed with:
```
Unhandled exception at 0x00007FFE3600B0B9 (ntdll.dll) in dotnet.exe.7132.dmp: 0xC0000374: A heap has been corrupted (parameters: 0x00007FFE360727F0).
```
![image](https://user-images.githubusercontent.com/4403806/84358312-4ece1300-ac0a-11ea-8b82-bb1cc24664c9.png)


Not quite sure what to look for though...

**Comment by weltkante** [comment]
I'll do the same process again and hope this reduces to a single test, seems WebBrowser wasn't the only issue.

**Comment by hughbe** [comment]
I know that the web browser clsid is being used in AxHost tests?

**Comment by weltkante** [comment]
> After merging #3423, which built successfully, the master build has failed

Thats unfortunate, sounds like there is another source of corruption then. I extended the PR to cover AxHost mentioned above, but I've not seen AxHost tests running in parallel with the other ones, so there still may be something else than WebBrowser.

Unfortunately I could not yet get local test runs to crash on the current master branch.

> Not quite sure what to look for though...

You'll want to look at all the threads being named after a unit test class
* If its waiting for a lock to create a handle you check if thats the first handle access in the control flow of the test (usually easily spottable). If so then this test is most likely not the source of corruption since it has not yet performed any significant interop.
* Otherwise, if the thread is in the middle of executing a test then it may have recently corrupted memory.
* If the thread is waiting for a lock with `Xunit.StaFact.dll!Xunit.Sdk.WinFormsSynchronizationContextAdapter.Create()` in the call stack then it didn't start the test yet and is waiting for its turn
* If the thread is in `Xunit.StaFact.dll!Xunit.Sdk.WinFormsSynchronizationContextAdapter.PumpTill` then its usually terminated recently (you could verify by activating the stack frame and looking for whether the `task` state is RanToCompletion but it usually is because we have only very few async tests). These are also candidates for cause of corruption.

By classifying threads and picking (2) and (4) from the list above you have a set of candidates which may cause corruption. From there I've been removing most of the other tests and kept running the unit test in a loop until it produced more dumps.

Also feel free to share the link to the dump or share it online somewhere if you want me to check it

**Comment by RussKie** [comment]
We did a bit of scouting around with Jeremy looking through the code in this stack. I'll try to set up a step-through environment, but from a high level it looks like `ImageList_Create` creates an instance of `CImageList` to casts it via [`QueryInterface`](https://docs.microsoft.com/en-us/windows/win32/api/unknwn/nf-unknwn-iunknown-queryinterface(refiid_void)) (not quite sure what `riid` is), and then fails while release the instance with failure: `heap_failure_block_not_busy`, which (I think) means the block of memory isn't owned by the caller (anymore?).
I suppose it is plausible that there is some kind of a race condition in `ImageList` implementations (using `Interop.ComCtl32.ImageList.*`).

**Comment by weltkante** [comment]
@Russkie I looked at the dump and this time `AxHostTests` were running in parallel with `HtmlDocumentTests` (both have a started but unfinished test thread), so it likely may have been `WebBrowser` again that caused the race condition and memory corruption. ImageList might have just been the victim of having the memory corrupted. In this case the last PR merged should have helped.

> heap_failure_block_not_busy, which (I think) means the block of memory isn't owned by the caller (anymore?).

Maybe, but if you look at the stack trace its been cut off (no managed caller), meaning the stack has been stomped over. The pointer to the `CImageList` may have been overwritten by garbage. Memory corruption is more likely than ownership mismanagement.

**Comment by weltkante** [comment]
@marklio if you still can make use of a TTD trace, I've got one [here](https://1drv.ms/u/s!AuW3VU_s-jxTgeZuaV9i1KAaV6Bi-Q?e=fRBEf4) - created it on an azure VM from the repro I linked, probably antivirus problems which prevent TTD working locally

**Comment by hughbe** [comment]
Can still reproduce this in https://dev.azure.com/dnceng/public/_build/results?buildId=693597&view=ms.vss-test-web.build-test-results-tab&runId=21511456&resultId=101437&paneView=debug which is rebased

**Comment by RussKie** [comment]
I'll check crash dumps tomorrow

**Comment by weltkante** [comment]
I've done a few runs but can't reproduce locally. I've only been given access to one dump after the 2nd fix was merged, and that dump was slightly corrupted so mixed mode debugger couldn't attach. Feel free to send me other dumps or links if I should be looking into them.

**Comment by weltkante** [comment]
@marklio ping for the available TTD trace [above](https://github.com/dotnet/winforms/issues/3358#issuecomment-643175448) in case you missed it.

**Comment by marklio** [comment]
@weltkante Thanks for the ping. I DID miss it. I'll try to look at that today.

**Comment by RussKie** [comment]
> See https://dev.azure.com/dnceng/public/_build/results?buildId=693597&view=ms.vss-test-web.build-test-results-tab&runId=21511456&resultId=101437&paneView=debug which is rebased

No crash dumps for this one.
There's a test failure, which looks like a general instability:
```
    System.Windows.Forms.Tests.ImageCollectionTests.ImageCollection_AddStrip_InvokeWithHandle_Success(transparentColor: Color [A=18, R=52, G=86, B=120], value: Bitmap { Flags = 2, FrameDimensionsList = [7462dc86-6180-4c7e-8e3f-ee7333a7a483], Height = 16, HorizontalResolution = 96, Palette = ColorPalette { Entries = [...], Flags = 7209033 }, ... }, expectedCount: 2) [FAIL]
      Assert.Equal() Failure
      Expected: 2
      Actual:   0
      Stack Trace:
        F:\workspace\_work\1\s\src\System.Windows.Forms\tests\UnitTests\System\Windows\Forms\ImageList.ImageCollectionTests.cs(976,0): at System.Windows.Forms.Tests.ImageCollectionTests.ImageCollection_AddStrip_InvokeWithHandle_Success(Color transparentColor, Image value, Int32 expectedCount)
```

**Comment by weltkante** [comment]
> There's a test failure, which looks like a general instability

I think its still the same memory corruption, it makes no sense that a bitmap that always has the same content "sometimes" is processed differently and adds no images to the strip. Memory corruption doesn't always have to lead to a crash.

**Comment by RussKie** [comment]
Let's get https://github.com/dotnet/winforms/pull/3470 first, and see if we get any observable changes. One thing @JeremyKuhne mentioned that he observed some object attempting to access external while being GC'ed, which could also lead to memory corruptions.

**Comment by weltkante** [comment]
found something, but still not the breakthrough we are looking for:
- dumps seem generally no longer debuggable, even dumps taken locally from task manager, opened dotnet/runtime#38174
- PR #3366 regressed WebBrowser tests, adding a new test without `[Collection(Sequential)]` (crashes are locally reproducable)
- Fixing above regression I still cannot reproduce the ImageList related crashes/failures, but I was running on a restricted test set based on the latest dump we examined, so maybe I removed too many tests.

I believe you had crashes before before the PR was merged? Its only worth looking at dumps after merging #3429 and before merging #3366. Will prepare a new PR to fix the regression

**Comment by RussKie** [comment]
What's really interesting our internal CI are not suffering the same instability as our public CI:
![image](https://user-images.githubusercontent.com/4403806/85263958-4ea80000-b4b3-11ea-8734-e733e2064cd0.png)

The only difference I can think of is that in the internal CI we only run Release configuration.
@AdamYoblick @MattGal is there anything else that is vastly different? 
Totally shooting from the heap here, we are running debug and release builds on the same machine, aren't we? I wonder if there is something on the native side that isn't getting torn down correctly, and then fails for subsequent calls coming from a x86 process... Notice all failures happen to x86 tests, which are run after x64 finish
```
  Running tests: F:\workspace\_work\1\s\artifacts\bin\System.Windows.Forms.Tests\Debug\netcoreapp5.0\System.Windows.Forms.Tests.dll [netcoreapp5.0|x64]
  Tests succeeded: F:\workspace\_work\1\s\artifacts\bin\System.Windows.Forms.Tests\Debug\netcoreapp5.0\System.Windows.Forms.Tests.dll [netcoreapp5.0|x64]
  Running tests: F:\workspace\_work\1\s\artifacts\bin\System.Windows.Forms.Tests\Debug\netcoreapp5.0\System.Windows.Forms.Tests.dll [netcoreapp5.0|x86]
  Tests succeeded: F:\workspace\_work\1\s\artifacts\bin\System.Windows.Forms.Tests\Debug\netcoreapp5.0\System.Windows.Forms.Tests.dll [netcoreapp5.0|x86]
```

**Comment by weltkante** [comment]
> What's really interesting our internal CI are not suffering the same instability as our public CI:

Thats odd, WebBrowser corruption was not specific to CI.

However, chances for crashes have reduced significantly with each of the three PRs I made, because it reduced the number of WebBrowser tests running in parallel. If internal tests never failed, even before the first of the three PRs, I'd question if internal tests are recognizing crashes as failures correctly? Before the first PR failures were common enough to happen "from time to time" even locally.

> we only run Release configuration.

I've seen both Debug and Release crash due to WebBrowser corruption

> Notice all failures happen to x86 tests, which are run after x64 finish

Thats not true, I've also seen WebBrowser corruptions crash in x64, but much rarer and before the first of the three PRs. Maybe x86 just has a higher chance for corruption to crash because of the smaller address space.

**Comment by RussKie** [comment]
I suppose if the current hypothesis is the WebBrowser tests causing failures, we should be able to test it relatively easy by just commenting out all WebBrowser related tests...

**Comment by weltkante** [comment]
WebBrowser tests causing failures is no hypothesis, I've been able to identify and reproduce that (see link to repro project above in the discussion). It only happens when executing those tests in parallel. Skipping them of course works as well _if you can catch all_. The problem is some tests invoke WebBrowser by GUID so its not obvious they instantiate a WebBrowser COM object.

The question is what the other source of memory corruption is. The third PR still crashed with heap corruption so I either missed another kind of "hidden" usage of WebBrowser COM objects, or there is a second yet unidentified source of memory corruption which I couldn't reproduce locally so far.

The problem is that the recent regression in .NET Core dumps being unreadable by VS (dotnet/runtime#38174) means I can't extract the information I need from crash dumps (namely managed stack traces to tell the state of the individual tests).

---

TLDR the current state is
* known corruption in WebBrowser COM objects when executed on more than one UI thread at a time
  * known workaround of marking all tests involving this COM object as sequential
  * I have a standalone repro and TTD trace for people to look at
* unknown source of corruption meaning either more usages of said COM object or something else yet unidentified
* a recent regression in .NET Core dump handling means we can't easily extract managed stack traces from dumps

**Comment by weltkante** [comment]
I can repro another corruption locally with a reduced set of unit tests, will continue to investigate and try to slice off more tests.

One thing I already noticed is that not all ImageLists are disposed, opened #3485 and #3486 for this. I don't think this is the source for the corruption since I cannot repro any crashes when I randomly allocate ImageLists in a standalone app, but it can't hurt to fix them.

**Comment by weltkante** [comment]
good news, one more corruption isolated, this time in the Common Controls library. The [repository](https://github.com/weltkante/WebBrowserCorruptionTest) is updated. Note that I just picked some [arbitrary test case](https://github.com/weltkante/WebBrowserCorruptionTest/blob/master/CoreCorruptionTest2/Program.cs#L56-L67) which reproed nicely, this corruption may be triggered by other tests as well. There is a nice small [TTD trace](https://1drv.ms/u/s!AuW3VU_s-jxTgeZ-23LnjPgvyCA-NA?e=WsvDww) available along with the [build output](https://1drv.ms/u/s!AuW3VU_s-jxTgeZ9mjIeULAywP3iKQ?e=2nnhGL) in case having the PDB helps.

It would be great if someone with access to the Common Controls source and knowledge of TTD debugging could take a look, since I cannot figure out what exactly is triggering the corruption, there is a bunch of property switching happening. I know Common Controls sometimes have issues [like this](https://devblogs.microsoft.com/oldnewthing/20171128-00/?p=97475), could be something similar since there is switching of image lists going on.

If you are able to figure out what exactly is causing the bug then you may be able to tell which tests to skip or how to work around the bug. As it stands I don't have any suggestions which tests could cause it.

Also note this corruption repros in .NET Core 5.0 and 3.1 as well as Desktop Framework so its unlikely to be an interop regression.

**Comment by RussKie** [comment]
I had spent past several days trying to identify the cause, sadly the culprit is still eluding me. There is some reportable progress though.

The current implementation fails on this line (that we knew):
https://github.com/dotnet/winforms/blob/fdb4350682224a6a8989b9559165d24dbf0d4f28/src/System.Windows.Forms/src/System/Windows/Forms/ImageList.cs#L458
However it was unclear why. Having restructured the code (see #3526) it became apparent that every so often `ComCtl32.ImageList.Create(...)` returns `IntPtr.Zero`, that then would go into `NativeImageList` and result in a NPE/AVE.

I have run `CoreCorruptionTest2` project from @weltkante's [repro](https://github.com/weltkante/WebBrowserCorruptionTest) (thank you!). Before #3526 it would crash for me almost immediately every time, but after the restructure I could often run tests for tens of minutes (making it harder to debug, hah), though still consistently observe failures.

With @JeremyKuhne we've looked at the comctl32 code but we couldn't spot anything obvious that could lead to `IntPtr.Zero`. The code is optimised that made setting a breakpoint or stepping through the code in question impossible (other than a disassembly).
In a nutshell, there are two calls that we think may lead to the failures (with no clear signs of why or under what conditions):
1. create an instance of native image list, and
2. `instance->QueryInterface(IImageList2)`.

I'm really punching above my weight here. To be continued...

**Comment by weltkante** [comment]
@RussKie @JeremyKuhne the TTD trace indicates that native `CImageList` is used by multiple threads at the same time, messing up the refcount. Try checking the return value of `ComCtl32.ImageList.Destroy(_himl)` in `NativeImageList.Dispose` - I suspect you are destroying an ImageList which no longer exists. If handles are reused you happen to destroy a random other ImageList which currently is in use on another thread. According to a quick test `ImageList_Destroy` is returning `FALSE` long before any memory corruption / crashing starts to happen.

<details><summary>snippet</summary>

```csharp

namespace System.Windows.Forms
{
    public sealed partial class ImageList
    {
        internal class NativeImageList : IDisposable, IHandle
        {
            public void Dispose(bool disposing)
            {
                if (_himl != IntPtr.Zero)
                {
                    var result = ComCtl32.ImageList.Destroy(_himl);
                    Debug.Assert(result.IsTrue());
                    _himl = IntPtr.Zero;
                }
            }
        }
    }
}
```

</details>

**Comment by JeremyKuhne** [comment]
@russkie had added locking to `NativeImageList`, I'll look through it with him later.

**Comment by RussKie** [comment]
I've rebased #3526 on top of the latest master (branch: https://github.com/RussKie/winforms/tree/Solve_ImageList_woes-repro-2), and now I'm observing AVE but in a completely different place 🤯 

```
 	[Inline Frame] ntdll.dll!RtlpIncrementCriticalSectionContentionCount(_RTL_CRITICAL_SECTION *) Line 1030	C	Symbols loaded.
 	ntdll.dll!RtlpWaitOnCriticalSection(_RTL_CRITICAL_SECTION * CriticalSection=0x0c872de0, unsigned long OldLockCount=4294967294) Line 1568	C	Symbols loaded.
 	ntdll.dll!RtlpEnterCriticalSectionContended(_RTL_CRITICAL_SECTION * CriticalSection=0x0c872de0) Line 2318	C	Symbols loaded.
 	ntdll.dll!RtlEnterCriticalSection(_RTL_CRITICAL_SECTION * CriticalSection=0x0c872de0) Line 1923	C	Symbols loaded.
>	[Inline Frame] comctl32.dll!CImageList::_EnterCritical() Line 41	C++	Symbols loaded.
 	comctl32.dll!CImageList::GetIconSize(int * pcx=0x0b4dc4cc, int * pcy=0x0b4dc4d0) Line 1485	C++	Symbols loaded.
 	comctl32.dll!ImageList_GetIconSize(_IMAGELIST * himl=0x0c872dc8, int * pcx=0x0b4dc4cc, int * pcy=0x0b4dc4d0) Line 8027	C++	Symbols loaded.
 	comctl32.dll!CLVImageListManager::_OnSetStateImageList(CLVDrawState * pDrawState=0x0c89bee0, _IMAGELIST * himlNew=0x0c872dc8, _IMAGELIST * * phimlOld=0x0b4dc4fc, bool * pbImageSizeChanged=0x0b4dc503) Line 186	C++	Symbols loaded.
 	comctl32.dll!CLVImageListManager::OnSetImageList(short sWorkArea=-1, int iImageList=2, _IMAGELIST * himl=0x0c872dc8) Line 48	C++	Symbols loaded.
 	comctl32.dll!CListView::WndProc(HWND__ * hwnd=0x003111bc, unsigned int uMsg=4099, unsigned int wParam=2, long lParam=210185672) Line 5231	C++	Symbols loaded.
 	comctl32.dll!CListView::s_WndProc(HWND__ * hwnd=0x003111bc, unsigned int uMsg=4099, unsigned int wParam=2, long lParam=210185672) Line 3605	C++	Symbols loaded.
 	user32.dll!__InternalCallWinProc@20() Line 116	Unknown	Symbols loaded.
 	[Inline Frame] user32.dll!InternalCallWinProc(long(__stdcall*)(HWND__ *, unsigned int, unsigned int, long)) Line 87	C++	Symbols loaded.
 	user32.dll!UserCallWinProcCheckWow(_ACTIVATION_CONTEXT * pActCtx, void * pfn=0x759b50e0, HWND__ * hwnd=0x003111bc, _WM_VALUE msg=4099, unsigned int wParam=2, long lParam=210185672, void * fEnableLiteHooks=0, int) Line 279	C++	Symbols loaded.
 	user32.dll!CallWindowProcAorW(long(__stdcall*)(HWND__ *, unsigned int, unsigned int, long) pfn=0x759b50e0, HWND__ * hwnd=0x003111bc, _WM_VALUE message=4099, unsigned int wParam=2, long lParam=210185672, int bAnsi=0) Line 2843	C++	Symbols loaded.
 	user32.dll!CallWindowProcW(long(__stdcall*)(HWND__ *, unsigned int, unsigned int, long) pfn=0x759b50e0, HWND__ * hwnd=0x003111bc, unsigned int message=4099, unsigned int wParam=2, long lParam=210185672) Line 2864	C++	Symbols loaded.
 	[Managed to Native Transition]		Annotated Frame
 	System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.DefWndProc(ref System.Windows.Forms.Message m = {System.Windows.Forms.Message}) Line 515	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.DefWndProc(ref System.Windows.Forms.Message m = {System.Windows.Forms.Message}) Line 5036	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.WndProc(ref System.Windows.Forms.Message m = {System.Windows.Forms.Message}) Line 13094	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.ListView.WndProc(ref System.Windows.Forms.Message m = {System.Windows.Forms.Message}) Line 6737	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.OnMessage(ref System.Windows.Forms.Message m = {System.Windows.Forms.Message}) Line 67	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.WndProc(ref System.Windows.Forms.Message m = {System.Windows.Forms.Message}) Line 119	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.Callback(System.IntPtr hWnd = 0x003111bc, Interop.User32.WM msg = 4099, System.IntPtr wparam = 0x00000002, System.IntPtr lparam = 0x0c872dc8) Line 372	C#	Symbols loaded.
 	[Native to Managed Transition]		Annotated Frame
 	user32.dll!__InternalCallWinProc@20() Line 116	Unknown	Symbols loaded.
 	[Inline Frame] user32.dll!InternalCallWinProc(long(__stdcall*)(HWND__ *, unsigned int, unsigned int, long)) Line 87	C++	Symbols loaded.
 	user32.dll!UserCallWinProcCheckWow(_ACTIVATION_CONTEXT * pActCtx, void * pfn=0x009d3736, HWND__ * hwnd=0x003111bc, _WM_VALUE msg=4099, unsigned int wParam=2, long lParam=210185672, void * fEnableLiteHooks=0, int) Line 279	C++	Symbols loaded.
 	user32.dll!CallWindowProcAorW(long(__stdcall*)(HWND__ *, unsigned int, unsigned int, long) pfn=0x009d3736, HWND__ * hwnd=0x003111bc, _WM_VALUE message=4099, unsigned int wParam=2, long lParam=210185672, int bAnsi=0) Line 2843	C++	Symbols loaded.
 	user32.dll!CallWindowProcW(long(__stdcall*)(HWND__ *, unsigned int, unsigned int, long) pfn=0x009d3736, HWND__ * hwnd=0x003111bc, unsigned int message=4099, unsigned int wParam=2, long lParam=210185672) Line 2864	C++	Symbols loaded.
 	[Inline Frame] comctl32.dll!CallOriginalWndProc(HWND__ *) Line 518	C	Symbols loaded.
 	comctl32.dll!CallNextSubclassProc(_SUBCLASS_HEADER * pHeader, HWND__ * hWnd=0x003111bc, unsigned int uMsg=4099, unsigned int wParam=2, long lParam=210185672) Line 1590	C	Symbols loaded.
 	[Inline Frame] comctl32.dll!DefSubclassProc(HWND__ *) Line 1191	C	Symbols loaded.
 	[Inline Frame] comctl32.dll!TTSubclassProc(HWND__ *) Line 1330	C	Symbols loaded.
 	comctl32.dll!CallNextSubclassProc(_SUBCLASS_HEADER * pHeader, HWND__ * hWnd=0x003111bc, unsigned int uMsg=4099, unsigned int wParam=2, long lParam=210185672) Line 1590	C	Symbols loaded.
 	comctl32.dll!MasterSubclassProc(HWND__ * hWnd=0x003111bc, unsigned int uMsg=4099, unsigned int wParam=2, long lParam=210185672) Line 1355	C	Symbols loaded.
 	user32.dll!__InternalCallWinProc@20() Line 116	Unknown	Symbols loaded.
 	[Inline Frame] user32.dll!InternalCallWinProc(long(__stdcall*)(HWND__ *, unsigned int, unsigned int, long)) Line 87	C++	Symbols loaded.
 	user32.dll!UserCallWinProcCheckWow(_ACTIVATION_CONTEXT * pActCtx, void * pfn=0x759ca480, HWND__ * hwnd=0x003111bc, _WM_VALUE msg=4099, unsigned int wParam=2, long lParam=210185672, void * fEnableLiteHooks=1, int) Line 279	C++	Symbols loaded.
 	user32.dll!SendMessageWorker(tagWND * pwnd=0x03a1fa50, unsigned int message=4099, unsigned int wParam=2, long lParam=210185672, int fAnsi=0) Line 877	C++	Symbols loaded.
 	user32.dll!SendMessageW(HWND__ * hwnd=0x003111bc, unsigned int message=4099, unsigned int wParam=2, long lParam=210185672) Line 906	C++	Symbols loaded.
 	[Managed to Native Transition]		Annotated Frame
 	System.Windows.Forms.Primitives.dll!Interop.User32.SendMessageW(IHandle hWnd = {System.Windows.Forms.ListView}, Interop.User32.WM Msg = 4099, System.IntPtr wParam = 0x00000002, System.IntPtr lParam = 0x0c872dc8) Line 36	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.ListView.RealizeProperties() Line 4835	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.ListView.OnHandleCreated(System.EventArgs e = {System.EventArgs}) Line 4467	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.WmCreate(ref System.Windows.Forms.Message m = {System.Windows.Forms.Message}) Line 11899	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.WndProc(ref System.Windows.Forms.Message m = {System.Windows.Forms.Message}) Line 12873	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.ListView.WndProc(ref System.Windows.Forms.Message m = {System.Windows.Forms.Message}) Line 6737	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.OnMessage(ref System.Windows.Forms.Message m = {System.Windows.Forms.Message}) Line 67	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.WndProc(ref System.Windows.Forms.Message m = {System.Windows.Forms.Message}) Line 119	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.Callback(System.IntPtr hWnd = 0x003111bc, Interop.User32.WM msg = CREATE, System.IntPtr wparam = 0x00000000, System.IntPtr lparam = 0x0b4dddd0) Line 372	C#	Symbols loaded.
```

![image](https://user-images.githubusercontent.com/4403806/86206528-f4442900-bbaf-11ea-811b-6e655db6754e.png)

**Comment by weltkante** [comment]
> now I'm observing AVE but in a completely different place

Most likely irrelevant. What I was trying to explain above is that an ImageList gets destroyed prematurely, so you can randomly fail _anywhere_ in ImageList code depending on how the memory is reused.

We did some external discussion and testing, currently it looks like enabling CheckBox mode (sometimes?) takes ownership of the ImageList and deletes it while WinForms still keeps the handle around. There is no multithreading needed, the StateImageList becomes nonfunctional once CheckBox mode is turned on. Adding multithreading just significantly increases the chance of handle colissions with something happening on the secondary thread.

Figuring out the correct ownership rules around checkbox mode should fix some if not all the flakyness of ImageList.

**Comment by RussKie** [comment]
I have created a [simple repro](https://github.com/RussKie/ListViewCheckboxesStateImageList) that toggles `CheckBoxes` and `StateImageList` properties on a `ListView` control. And if a `CheckBoxes` is toggled _whilst_ a user-supplied imagelist is bound to `StateImageList` property - the native listview control will tarnish the user-supplied one. 
The issue is reproducible in net4x and netcore3.1.

**Comment by RussKie** [comment]
Observed yet another heap corruption linked to `ImageList` use in https://dev.azure.com/dnceng/public/_build/results?buildId=727355&view=logs&jobId=1f132584-ab08-5fbf-e5d0-5685de454342 for https://github.com/dotnet/winforms/pull/3569

![image](https://user-images.githubusercontent.com/4403806/87377013-38cbbd80-c5cf-11ea-899f-bed9112edc5b.png)


[Memory dump](https://helixri107v0xdeko0k025g8.blob.core.windows.net/results-aa22050bc8c44ba994/928623f1-8b1d-4642-8519-afb9ac7e913e/dotnet.exe.412.dmp?sv=2019-02-02&se=2020-10-11T18%3A16%3A23Z&sr=c&sp=rl&sig=GIE9DEl3XG%2FdHmssHIe2hBh%2FgnT4IpurgBhzampenuw%3D)

**Comment by weltkante** [comment]
I actually can't see PR #3526 addressing ListView ImageList ownership when switching on checked mode, so I'd guess its still the same corruption. Initially I had assumed it was just moved over to #3531 but considering that is 'Future' now I guess there has been some confusion.

**Comment by RussKie** [comment]
Yes, we had hopes that #3526 would help, but obviously we're wrong. We'll have to continue digging...

**Comment by hughbe** [comment]
Did that PR help? It seems like there are like 5 different problems here - web browser, disposal, sharing of image lists, etc

**Comment by weltkante** [comment]
To me it looks pretty clear. While there were a lot of secondary issues found while investigating there is only two sources of crashes identified:
- WebBrowser when running on more than one UI thread at a time (has been "sequentialized" so should no longer trigger)
- ListView when having a custom ImageList and turning Checked on (as far as I can tell this has a follow-up issue #3531 but has not been acted upon yet)

Everything else has been split off into separate issues and has not been the source of crashes seen here.

Maybe it was not clear that turning on Checkboxes causes the crashes? It will trigger a double-free of the underlying native ImageList because there is a small chance that the same numeric handle value is reused again for an independent ImageList.

**Comment by RussKie** [comment]
I don't believe WebBrowser has anything to do with these failures. I haven't observed it in any of the dumps I have.

Here's one of the latests AV crashes:
```
Not Flagged	>	42280	0	Worker Thread	System.Windows.Forms.Tests.ListViewTests.ListView_StateImageList_SetWithHandleWithNonNullOldValue_GetReturnsExpected	win32u.dll!_NtUserWaitMessage@0
Not Flagged		38412	0	Worker Thread	System.Windows.Forms.Tests.ScrollableControlTests.ScrollableControl_OnPaintBackground_InvokeWithHandle_Success	win32u.dll!_NtUserMsgWaitForMultipleObjectsEx@20
Not Flagged		40560	0	Worker Thread	System.Windows.Forms.Tests.ControlTests.Control_IsInputChar_InvokeWithHandle_ReturnsExpected	win32u.dll!_NtUserMsgWaitForMultipleObjectsEx@20
Not Flagged		42088	0	Worker Thread	System.Windows.Forms.Tests.TextBoxBaseTests.TextBoxBase_WndProc_InvokeMouseDownWithoutHandle_Success	win32u.dll!_NtUserMsgWaitForMultipleObjectsEx@20
Not Flagged		39268	0	Worker Thread	System.Windows.Forms.Tests.ToolStripItemTests.ToolStripItem_ImageAlign_SetWithOwnerWithHandle_GetReturnsExpected	win32u.dll!_NtUserMsgWaitForMultipleObjectsEx@20
Not Flagged		40208	0	Worker Thread	System.Windows.Forms.Tests.ButtonBaseTests.ButtonBase_OnMouseMove_InvokeWithHandle_CallsMouseMove	win32u.dll!_NtUserMsgWaitForMultipleObjectsEx@20
Not Flagged		42608	0	Worker Thread	System.Windows.Forms.Tests.TabPageTests.TabPageCollection_Text_SetGetItemsWithHandleDesignMode_Success	win32u.dll!_NtUserMsgWaitForMultipleObjectsEx@20
Not Flagged		48308	0	Worker Thread	System.Windows.Forms.Tests.ControlControlCollectionTests.ControlCollection_Add_ControlNewCollection_Success	win32u.dll!_NtUserMsgWaitForMultipleObjectsEx@20
Not Flagged		40188	0	Worker Thread	System.Windows.Forms.Tests.RichTextBoxTests.RichTextBox_Find_CharArrayIntInt_ReturnsExpected	win32u.dll!_NtUserMsgWaitForMultipleObjectsEx@20
Not Flagged		41032	0	Worker Thread	System.Windows.Forms.Tests.TreeViewTests.TreeView_OnHandleCreated_InvokeWithHandleWithProperties_CallsHandleCreated	win32u.dll!_NtUserMsgWaitForMultipleObjectsEx@20
Not Flagged		42672	0	Worker Thread	System.Windows.Forms.Tests.ToolStripTests.ToolStrip_OnControlAdded_Invoke_CallsControlAdded	win32u.dll!_NtUserMsgWaitForMultipleObjectsEx@20
```

This crash also results in a popup that hangs builds:
![image](https://user-images.githubusercontent.com/4403806/87642771-4dee4b00-c78d-11ea-868e-aac82f9f3c6e.png)



> * ListView when having a custom ImageList and turning checked mode on (as far as I can tell this has a follow-up issue #3531 but has not been acted upon yet)
>
> Maybe it was not clear that turning on Checkboxes causes the crashes? It will trigger a double-free of the underlying native ImageList because there is a small chance that the same numeric handle value is reused again for an independent ImageList.
> 
> Fixing this requires handling ImageList ownership correctly when turning checkbox mode on, the native code is assuming it owns the ImageList and releases it, WinForms must not retain its copy of the handle. Since you don't control who else is using it you have to duplicate it before giving it to the ListView. Alternatively it might be possible to work around this by resetting the ImageList on the native control before turning checkbox mode on.

As far as our research went this is the most likely cause, but so far we were unable to get to the bottom of the issue...
The code will require a refactor and a rewrite, and this requires time.

**Comment by weltkante** [comment]
> I don't believe WebBrowser has anything to do with these failures. I haven't observed it in any of the dumps I have.

They are reproducible though, even without any ImageList in play, see my other repro scenario. Its not always easy to separate dumps caused from memory corruption, in hindsight you may be right and the CI dumps are probably mostly ImageList handle collisions. The WebBrowser and ImageList tests always happened to run at roughly the same time and corruption from WebBrowser was much more frequent, at least for me running tests locally, so it reproduced more easily and was discovered first.

I don't think there is any point separating them into an issue of their own considering the WinForms side of the investigation is already over and the workaround in place, but if you want to track the response of the external investigation so you can reconsider removing the sequential attributes from tests I can create a separate issue if you want.

**Comment by weltkante** [comment]
> Fixing this requires handling ImageList ownership correctly when turning checkbox mode on

Made a PR which disables ImageList ownership sharing and instead manages ownership explicitly by creating unique ImageList handles when assigning to the native control. Your choice of whether you want to take it or live with the broken tests for now (until you can fix it), I've no strong opinion, I don't think it happens much in practice outside the unit tests

**Comment by hughbe** [comment]
It would be really interesting to see if there’s some report of this snuck away in stack overflow or forums somewhere

**Comment by RussKie** [comment]
For general reference, stumbled across this post that alleges an imagelist may be double destroyed if it is still bound to a control
https://social.msdn.microsoft.com/Forums/vstudio/en-US/e6b685d3-e407-4d5a-a7b7-faa0f7840375/cimagelistrelease-mfc-vs2005-winxp

**Comment by weltkante** [comment]
I wouldn't put too much weight on that, he just saw the same thing I did. In particular its no indication of where the bug is, without repro to follow along you can't tell who is _responsible_ for the off-by-one refcount. The incorrect release could have happened anywhere, it will just manifest in a bug when the _last_ reference is released, so it will almost always blame the wrong location in the callstack, as the refcount could have been off-by-one for a long time.
